Public/Src/Cache/ContentStore/Interfaces/Synchronization/SemaphoreSlimToken.cs

```diff
@@ -42,6 +42,20 @@ public static SemaphoreSlimToken Wait(SemaphoreSlim semaphore)
             return new SemaphoreSlimToken(semaphore);
         }
 
+        /// <summary>
+        ///     Wait on a SemaphoreSlim and return a token that, when disposed, calls Release() on the SemaphoreSlim.
+        ///     If the deadline is reached, the token is not acquired and a fake disposable is returned.
+        /// </summary>
+        /// <param name="semaphore">The semaphore to wait on</param>
+        /// <param name="millisecondsTimeout">Maximum time to wait for the semaphore, in milliseconds</param>
+        /// <param name="acquired">Whether the semaphore was obtained or not</param>
+        /// <returns>A token that, when disposed, calls Release() on the SemaphoreSlim</returns>
+        public static SemaphoreSlimToken TryWait(SemaphoreSlim semaphore, int millisecondsTimeout, out bool acquired)

```

Comment:
2 very nit comments (feel free to ignore):

1. Move this to SemaphoreSlimExtensions to simplify usage
2. Add an overload without timeout (will pass 0 by default)
3. use new 'using var' in ConentLocationDatabase (where this extension is used) to remove extra block:

using var _gate.TryWait(out var acquired);